scipy/sparse/coo.py

```diff
@@ -205,10 +205,17 @@ def reshape(self, *args, **kwargs):
         nrows, ncols = self.shape
 
         if order == 'C':
-            flat_indices = ncols * self.row + self.col
+            # Upcast to avoid overflows: the coo_matrix constructor
+            # below will downcast the results to a smaller dtype, if
+            # possible.
+            dtype = get_index_dtype(maxval=(ncols * max(0, nrows - 1) + max(0, ncols - 1)))
+
+            flat_indices = np.multiply(ncols, self.row, dtype=dtype) + self.col
             new_row, new_col = divmod(flat_indices, shape[1])
         elif order == 'F':
-            flat_indices = self.row + nrows * self.col
+            dtype = get_index_dtype(maxval=(nrows * max(0, ncols - 1) + max(0, nrows - 1)))
+
+            flat_indices = np.multiply(nrows, self.col, dtype=dtype) + self.row
             new_col, new_row = divmod(flat_indices, shape[0])
```

Comment:
I forget what the minimum numpy version is, but if it's >=1.6.0 we can do:

```
new_row, new_col = np.unravel_index(flat_indices, shape, order)
```

instead of custom `divmod` for each branch.

Reply:
It's 1.8.2, https://github.com/scipy/scipy/blob/master/setup.py#L409

Reply:
But I guess it can't be used here, because the dtype choice for `flat_indices` needs to depend on dtype, and that depends on `order`.